clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/MemoryEventRepository.kt

```diff
@@ -0,0 +1,168 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.clouddriver.event.persistence
+
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.clouddriver.event.Aggregate
+import com.netflix.spinnaker.clouddriver.event.EventMetadata
+import com.netflix.spinnaker.clouddriver.event.EventPublisher
+import com.netflix.spinnaker.clouddriver.event.SpinEvent
+import com.netflix.spinnaker.clouddriver.event.config.MemoryEventRepositoryConfigProperties
+import com.netflix.spinnaker.clouddriver.event.exceptions.AggregateChangeRejectedException
+import com.netflix.spinnaker.kork.exceptions.SystemException
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.math.max
+
+/**
+ * An in-memory only [EventRepository]. This implementation should only be used for testing.
+ */
+class MemoryEventRepository(
+  private val config: MemoryEventRepositoryConfigProperties,
+  private val eventPublisher: EventPublisher,
+  private val registry: Registry
+) : EventRepository {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val aggregateCountId = registry.createId("eventing.aggregates")
+  private val aggregateWriteCountId = registry.createId("eventing.aggregates.writes")
+  private val aggregateReadCountId = registry.createId("eventing.aggregates.reads")
+  private val eventCountId = registry.createId("eventing.events")
+  private val eventWriteCountId = registry.createId("eventing.events.writes")
+  private val eventReadCountId = registry.createId("eventing.events.reads")
+
+  private val events: MutableMap<Aggregate, MutableList<SpinEvent>> = ConcurrentHashMap()
+
+  override fun save(aggregateType: String, aggregateId: String, originatingVersion: Long, events: List<SpinEvent>) {
+    registry.counter(aggregateWriteCountId).increment()
+
+    val aggregate = getAggregate(aggregateType, aggregateId)
+
+    if (aggregate.version != originatingVersion) {
+      // If this is being thrown, ensure that the originating process is retried on the latest aggregate version
+      // by re-reading the events list.
+      throw AggregateChangeRejectedException(
+        "Attempting to save events against an old aggregate version " +
+        "(version: ${aggregate.version}, originatingVersion: $originatingVersion)")
+    }
+
+    val currentSequence = this.events[aggregate]!!.map { it.metadata.sequence }.max() ?: 0
```

Comment:
unless i am missing something, i don't think the sequence calculation is thread-safe

Reply:
actually also the sequence assignment on L69 is not safe either unless this whole method is in a critical section (or equivalent)

Reply:
My assumption was that an aggregate would be written to by a single thread. Probably a poor assumption, so I'll update.

Reply:
ah, yeah, that makes sense - didn't realize it was meant that only 1 thread per aggregate